### DISCUSSION TITLE
Suggestion: debug mode

### DISCUSSION THREAD

**Discussion by jacksondc** [yasserfarouk/scml#80] [Ideas]
Is it possible to create a debug mode, where you could pause a simulation and then walk through step by step?

**Comment by yasserfarouk** [comment]
If you are running a single world, that is easily accomplished by simply calling `step()` instead of `run()` on the world. Something like:

```python
world = SCMLOneshotWorld(...)
for s in range(world.n_steps):
    world.step()
```

Moreover, NegMAS supports an event listener model in which you can register your own listener on different events. There are three ways to do that:

### Stats monitor:

```python
class MyMonitor(StatsMonitor):
    """A monitor object capable of receiving stats of a world"""

    def init(self, stats: Dict[str, Any], world_name: str):
        """Called to initialize the monitor before running first step"""

    def step(self, stats: Dict[str, Any], world_name: str):
        """Called at the END of every simulation step"""
```

Then you can create the world and pass it the stats monitor:

```python
world = SCMLOneshotWorld(..., stats_monitors=[MyMonitor()], ...)
```

This allows you to monitor statistics but nothing more

### World monitor

class MyMonitor(WorldMonitor):
    """A monitor object capable of monitoring a world. It has read/write access to the world"""

    def init(self, world: "World"):
        """Called to initialize the monitor before running first step"""

    def step(self, world: "World"):
        """Called at the END of every simulation step"""

Then you can create the world and pass it the stats monitor:

```python
world = SCMLOneshotWorld(..., world_monitors=[MyMonitor()], ...)
```

This allows you to receive a callback at the end of every simulation step and you can do whatever you want including breaking into a debugger.


### General NegMAS event system

This is the most advanced version. It allows you to monitor any specific event raised by the system.
You can find a list of all events raised by world simulations [here](http://yasserm.com/negmas/_modules/negmas/situated/situated.html)

```python
class MyMonitor(EventSink):
        def on_event(self, event: Event, sender: EventSource):
               """Will be received whenever and avent is raised"""
```

To use this version, you need to register your monitor using `register_listener()` after the world is created passing it the name of the event you want to listen to.

This can be helpful in debugging when a specific event is bothering you and it is hard to find out when exactly does it happen.
